### PR TITLE
Add ExternalAccess for Xamarin

### PR DESCRIPTION
**Pull request by kzu**
Xamarin is planning on unifying some Rosln-powered analyzers, code
fixers and completion providers into a new shared (for both VS and
VSfM) repository and some of the existing code that currently lives
in MonoDevelop repo requires internal access.

### CODE REVIEW

**Comment by kzu** [comment]
Will wait for #34049 to be merged before rebasing, since that will cause conflicts on the `.sln` again anyway.

**Comment by kzu** [comment]
Rebased on master

**Comment by kzu** [comment]
Gotta love all those green checks @sharwell 👍

**Comment by kzu** [comment]
Feedback addressed :)

File: src/Workspaces/Core/Desktop/Microsoft.CodeAnalysis.Workspaces.Desktop.csproj [hunk 1]
```diff
@@ -30,6 +30,7 @@
     <InternalsVisibleTo Include="Roslyn.VisualStudio.Setup" />
     <InternalsVisibleTo Include="Roslyn.VisualStudio.DiagnosticsWindow" />
     <InternalsVisibleTo Include="Roslyn.VisualStudio.RemoteHostClientMock" />
+    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.Xamarin" />
```
[the remark this answers is not in the document]
Comment (kzu) [reply src/Workspaces/Core/Desktop/Microsoft.CodeAnalysis.Workspaces.Desktop.csproj]: Removed this change

File: src/NuGet/VisualStudio/VS.ExternalAPIs.Roslyn.Package.csproj [hunk 2]
```diff
@@ -42,6 +42,8 @@
     <ProjectReference Include="..\..\Tools\ExternalAccess\Razor\Microsoft.CodeAnalysis.ExternalAccess.Razor.csproj" PrivateAssets="all" />
     <ProjectReference Include="..\..\Tools\ExternalAccess\TypeScript\Microsoft.CodeAnalysis.ExternalAccess.TypeScript.csproj" PrivateAssets="all" />
     <ProjectReference Include="..\..\Tools\ExternalAccess\UnitTesting\Microsoft.CodeAnalysis.ExternalAccess.UnitTesting.csproj" PrivateAssets="all" />
+    <ProjectReference Include="..\..\Tools\ExternalAccess\Xamarin\Microsoft.CodeAnalysis.ExternalAccess.Xamarin.csproj" PrivateAssets="all" />
+    <ProjectReference Include="..\..\Tools\ExternalAccess\Xamarin.Remote\Microsoft.CodeAnalysis.ExternalAccess.Xamarin.Remote.csproj" PrivateAssets="all" />
```
Comment (sharwell) [inline src/NuGet/VisualStudio/VS.ExternalAPIs.Roslyn.Package.csproj]: 📝 These should also be part of a pair of `_File` elements below

Comment (kzu) [reply src/NuGet/VisualStudio/VS.ExternalAPIs.Roslyn.Package.csproj]: Added

**Comment by kzu** [comment]
Rebased with master and fixed sln conflicts

**Comment by kzu** [comment]
Rebased, again, fixed conflicts...

**Comment by sharwell** [comment]
@kzu we were not able to get the assembly loads approved as part of the new external access assemblies. Rather than place your external access code here, you'll want to place the exact same code in the original roslyn assemblies as seen in https://github.com/dotnet/roslyn/pull/36151.

**Comment by kzu** [comment]
So, to make sure I understood correctly, the steps would be:

1. Delete the entire `Microsoft.CodeAnalysis.ExternalAccess.Xamarin` project and all its references from everywhere I added it before (project references, solution, etc.
2. Replace all the existing `<InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.Xamarin" />` with `<RestrictedInternalsVisibleTo Include="Xamarin.CodeAnalysis" Key="$(XamarinKey)" />` (that is, in the specific projects I need it, such as `Microsoft.CodeAnalysis.EditorFeatures.csproj` and such)

And that's it?

**Comment by sharwell** [comment]
@kzu Basically yes. You'll also want to set `Partner="Xamarin"` on the `RestrictedInternalsVisibleTo` line.

We are currently validating the new workflow in 16.3 Preview 1 and will be ready for other teams to follow suit as soon as we can get #36151 merged.

**Comment by kzu** [comment]
Unless I'm reading it wrong, if I add `Partner=Xamarin`, I'll get a [restricted namespce](https://github.com/dotnet/roslyn/pull/36151/files#diff-72479b7e777c44372d1495e9765309cbR248) for the `InternalsVisibleTo` from  from the [RestrictedInternalsVisibleTo](https://github.com/dotnet/roslyn/pull/36151/files#diff-0e3b4dcf2402e0c902fd3523543d4a71R10). I'd rather just keep the IVT to the assembly, not specific namespace...

**Comment by kzu** [comment]
(also, I meant delete project from this branch/PR, since I'm going to keep tweaking it ;))

**Comment by sharwell** [comment]
@kzu External IVTs are required to have restricted accessibility. You can define the types you need in the assigned namespace and must not reference internal types outside that namespace. The restricted IVT analyzer will enforce this separation.

**Comment by kzu** [comment]
I may be missing something. Does this mean that we'll now have to go over *every* internal roslyn API used in these IVTs and create "accessors" like the PR you pointed to for each and every one?

```
    <InternalsVisibleTo Include="MonoDevelop.Ide" Key="$(MonoDevelopKey)" LoadsWithinVisualStudio="false" />
    <InternalsVisibleTo Include="MonoDevelop.Refactoring" Key="$(MonoDevelopKey)" LoadsWithinVisualStudio="false" />
    <InternalsVisibleTo Include="MonoDevelop.CSharpBinding" Key="$(MonoDevelopKey)" LoadsWithinVisualStudio="false" />
    <InternalsVisibleTo Include="MonoDevelop.VBNetBinding" Key="$(MonoDevelopKey)" LoadsWithinVisualStudio="false" />
    <InternalsVisibleTo Include="MonoDevelop.Ide.Tests" Key="$(MonoDevelopKey)" LoadsWithinVisualStudio="false" />
    <InternalsVisibleTo Include="MonoDevelop.Refactoring.Tests" Key="$(MonoDevelopKey)" LoadsWithinVisualStudio="false" />
    <InternalsVisibleTo Include="MonoDevelop.CSharpBinding.Tests" Key="$(MonoDevelopKey)" LoadsWithinVisualStudio="false" />
    <InternalsVisibleTo Include="MonoDevelop.VBNetBinding.Tests" Key="$(MonoDevelopKey)" LoadsWithinVisualStudio="false" />
```

(from https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/Core/Microsoft.CodeAnalysis.EditorFeatures.csproj#L59-L66 and other projects are similar). 

Because what I'm trying to achieve here is being able to move some of those analyers that aren't MonoDevelop/VSM-specific and are instead *Xamarin*-specific over to a new assembly/repo where we can share them with VS. I was hoping it was as easy as adding our new assembly to that list, but this is looking more and more complicated every time we iterate on this :(

**Comment by sharwell** [comment]
@kzu No, you only need to make those changes for assemblies that ship with or are known to load within Visual Studio.

**Comment by kzu** [comment]
Maybe I'm not expressing myself correctly here, sorry :(. _some_ of the types (analyzers/completion/what-not) _will_ start moving to a new `Xamarin.CodeAnalysis` assembly which _will_ load both in VS *and* VSM (since we want to start sharing more code across both IDEs for Xamarin). So my question is: what's the policy for that?

**Comment by sharwell** [comment]
The basic idea is correct. However, it likely won't be as much work as you imagine. We have a new analyzer for restricted IVTs that does all the enforcement for you, so the manual audit is unnecessary. I might be able to take things one step further and create a refactoring that automatically creates the necessary shims (once the patterns stabilize).

**Comment by kzu** [comment]
Ok, got a quick call, made progress, I'll open a new one once we get further along this path.

**Pull request closed without merging**